src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs

```diff
@@ -316,21 +316,27 @@ private void MoveNext(Thread? threadPoolThread)
                     AsyncCausalityTracer.TraceSynchronousWorkStart(this, CausalitySynchronousWork.Execution);
                 }
 
+                Debug.Assert(StateMachine != null);
+
                 ExecutionContext? context = Context;
                 if (context == null)
                 {
-                    Debug.Assert(StateMachine != null);
                     StateMachine.MoveNext();
                 }
+                else if (threadPoolThread is null)
+                {
+                    ExecutionContext.RunInternal(context, s_callback, this);
+                }
                 else
                 {
-                    if (threadPoolThread is null)
+                    if (context.IsDefault)
                     {
-                        ExecutionContext.RunInternal(context, s_callback, this);
+                        // Context starts as Default on ThreadPool
+                        StateMachine.MoveNext();
```

Comment:
I'm not following why this is valid.  MoveNext could be invoked, for example, as part of invoking task continuations; if StateMachine.MoveNext makes EC changes, what resets them in such cases?

Reply:
If `threadPoolThread` is set then its calling it directly from `ThreadPool.Dispatch`; it uses `ExecutionContext.RunInternal` (above) when its `null`